### PR TITLE
Logic to convert both GST and CPR datasets to Hugginface datasets so …

### PR DESCRIPTION
**Pull request by stefl14**
…that we can use to elastic search, to faiss etc.

### CODE REVIEW

File: src/cpr_data_access/models.py [hunk 1]
```diff
@@ -686,3 +693,71 @@ def get_all_text_blocks(
...
+        return flattened_docs
+
+    def to_huggingface(self) -> HFDataset:
+        """
+        Convert to a huggingface dataset to get access to the huggingface datasets API.
+
+        :return: Huggingface dataset
+        """
+        flattened_docs = self._flatten_structure()
+        hf_dataset = HFDataset.from_dict({k: [d[k] for d in flattened_docs] for k in flattened_docs[0]})
```
Comment (kdutia) [inline src/cpr_data_access/models.py:762]: This line was quite difficult to understand. I think it's worth renaming the variables `d` and `k` (e.g. `document`, `document_keys`) and assigning `document_keys = flattened_docs[0].keys()` explicitly on a separate line.

File: tests/test_models.py [hunk 2]
```diff
@@ -276,3 +285,23 @@ def test_document_get_text_window(test_document):
...
+    assert list(dataset_hf.features.keys()) == ['text_block_id', 'text', 'type', 'language', 'type_confidence',
+                                                'page_number', 'coords', 'document_id', 'document_name',
+                                                'document_source_url', 'document_content_type', 'document_md5_sum',
+                                                'languages', 'translated', 'has_valid_text', 'publication_ts',
+                                                'geography', 'category', 'source', 'sectors', 'document_url',
+                                                'document_cdn_object', 'document_description', 'document_slug']
+    assert list(dataset_gst_hf.features.keys()) == ['text_block_id', 'text', 'type', 'language', 'type_confidence',
+                                                    'page_number', 'coords', 'document_id', 'document_name',
+                                                    'document_source_url', 'document_content_type', 'document_md5_sum',
+                                                    'languages', 'translated', 'has_valid_text', 'geography']
```
Comment (kdutia) [inline tests/test_models.py:307]: you might be able to check against `model.__fields__` here to make this more dynamic against future pydantic model changes

File: src/cpr_data_access/models.py [hunk 3]
```diff
@@ -537,6 +540,10 @@ class CPRDocumentWithURL(CPRDocument):
     document_url: Optional[AnyHttpUrl]
 
 
+class DataNotLoadedException(Exception):
+    pass
+
+
```
Comment (kdutia) [inline src/cpr_data_access/models.py:546]: this appears not to be used?

File: src/cpr_data_access/models.py [hunk 4]
```diff
@@ -686,3 +693,71 @@ def get_all_text_blocks(
                         output_values.append(block)
 
         return output_values
+
+    def _create_text_block_doc(self, doc: AnyDocument, block: TextBlock) -> dict[str, Any]:
```
Comment (kdutia) [inline src/cpr_data_access/models.py:697]: This should live outside of the `Dataset` class and I think a name like `_create_text_block_and_document_dict` might be clearer

File: src/cpr_data_access/models.py [hunk 5]
```diff
@@ -686,3 +693,71 @@ def get_all_text_blocks(
...
+
+        if self.document_model == CPRDocument:
+            flattened_doc['document_url'] = doc.document_url
+            flattened_doc['document_cdn_object'] = doc.document_cdn_object
+            flattened_doc['document_description'] = doc.document_description
+            flattened_doc['document_slug'] = doc.document_slug
+
+        return flattened_doc
+
+    def _flatten_structure(self) -> List[dict[str, Any]]:
```
Comment (kdutia) [inline src/cpr_data_access/models.py:739]: `_to_list_of_dicts`?

File: pyproject.toml [hunk 6]
```diff
@@ -13,6 +13,7 @@ boto3 = "^1.26.16"
 tqdm = "^4.64.1"
 aws-error-utils = "^2.7.0"
 pandas = "^1.5.3"
+datasets = "^2.10.1"
```
Comment (kdutia) [inline pyproject.toml:16]: this is an interesting decision point should we expand the data access lib to use features from more libraries – should we install the libraries specifically, or warn if the user hasn't got the library installed?

Comment (stefl14) [reply pyproject.toml:16]: hmm not sure, don't have a strong opinion.

File: src/cpr_data_access/models.py (hunk 3 above)
Comment (stefl14) [reply src/cpr_data_access/models.py]: Ah yes, decided it was unnecessary so remove dit.

File: tests/test_models.py (hunk 2 above)
Comment (kdutia) [reply tests/test_models.py]: as in just checking the keys? think the other assert statements are still valuable

**Review by kdutia — APPROVED** [review]
nice!

**Pull request merged**